crates/metal-codegen-llvm/src/safeties.rs

```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: MIT
+//! Safe Rust abstractions for several LLVM types
+
+use std::slice;
+
+use llvm_sys::{
+    core::{LLVMDisposeMemoryBuffer, LLVMGetBufferSize, LLVMGetBufferStart},
+    prelude::LLVMMemoryBufferRef,
+};
+
+/// Safe abstraction over LLVM's MemoryBuffer
+pub struct MemoryBuffer {
+    llvm_buf: LLVMMemoryBufferRef,
+}
+
+impl MemoryBuffer {
+    pub fn new(buf: LLVMMemoryBufferRef) -> Self {
+        Self { llvm_buf: buf }
+    }
+
+    /// Turns an LLVM buffer into a Vec<u8>
+    #[inline]
+    pub fn to_vec(&self) -> Vec<u8> {
+        unsafe {
+            let start = LLVMGetBufferStart(self.llvm_buf) as *const u8;
+            let size = LLVMGetBufferSize(self.llvm_buf) as usize;
+            slice::from_raw_parts(start, size).to_vec()
+        }
+    }
```

Comment:
Wouldn't it be better to return a &[u8]? Rust should even automatically infer that it will only live as long as `self`

Reply:
> Wouldn't it be better to return a &[u8]? Rust should even automatically infer that it will only live as long as `self`

It's supposed to live longer than `self`, especially since it's returned in `compile_module`.

Reply:
The *slice* is not supposed to live longer than self, and if the caller wants an owned vec, it's simple to add a `.to_vec()` call

Reply:
> The _slice_ is not supposed to live longer than self, and if the caller wants an owned vec, it's simple to add a `.to_vec()` call

this function is `to_vec` by itself though. If it was to bytes or to slice definitely, but this function should return an owned vec and that's it, no need to add more steps.